### PR TITLE
[TimexExpression] Support merged timex of duration/datetimerange like "PT1H30M" (#1515, #1530, #1923, #2110)

### PR DESCRIPTION
**Pull request by ChopperMan33**
Support merged timex of duration/datetimerange like "PT1H30M", "(2020-06-04T15,2020-06-04T17:30,PT2H30M)".(#1515, #1923, #2110)
Fix time addition of start/end in duration/timerange. (#1530)
Add constants of TimexExpression.
Support Invariant Culture in parsing.
Add unit test.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexRegex.cs [hunk 1]
```diff
@@ -52,14 +53,16 @@ public static class TimexRegex
                 PeriodCollectionName, new Regex[]
                 {
                     new Regex(@"^P(?<amount>\d*\.?\d+)(?<dateUnit>Y|M|W|D)$"),
-                    new Regex(@"^PT(?<amount>\d*\.?\d+)(?<timeUnit>H|M|S)$"),
+                    new Regex(@"^PT(?<hourAmount>\d*\.?\d+)H(\d*\.?\d+(M|S)){0,2}$"),
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexRegex.cs:56]: Do we need these to be separate regexes? My worry here is performance.
If combining would make readability much worse, don't do it.

**Review by tellarin — CHANGES_REQUESTED** [review]

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs [hunk 2]
```diff
@@ -63,42 +64,47 @@ public static string Format(TimexProperty timex)
 
         private static string FormatDuration(TimexProperty timex)
         {
+            var isTimeDurationAlreadyExist = false;
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs:67]: It is used for showing whether we need to add "T" when we append "T30M", "T30S" in timex. If the timex already contains a "T", it should only append "30M", "30s“.
refer to: 
    https://github.com/microsoft/Recognizers-Text/blob/master/.NET/Microsoft.Recognizers.Text.DateTime/Utilities/TimexUtility.cs#L56

File: .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs [hunk 3]
```diff
@@ -338,6 +351,62 @@ public void DataTypes_Resolver_DateRange_Demaical_Period_EN()
             CultureInfo.CurrentCulture = sourceLanguage;
         }
 
+        [TestMethod]
+        public void DataTypes_Resolver_TimeRange_11_30_to_12_00()
+        {
+            var today = System.DateTime.Now;
+            var resolution = TimexResolver.Resolve(new[] { "(T11:30,T12:00,PT30M)" }, today);
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs:358]: Yes. I will add a test case for it.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs [hunk 4]
```diff
@@ -125,12 +131,12 @@ private static string FormatDate(TimexProperty timex)
...
             if (timex.Month != null && timex.DayOfMonth != null)
             {
-                return $"XXXX-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-{TimexDateHelpers.FixedFormatNumber(timex.DayOfMonth, 2)}";
+                return $"{Constants.TimexFuzzyYear}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-{TimexDateHelpers.FixedFormatNumber(timex.DayOfMonth, 2)}";
             }
 
             if (timex.DayOfWeek != null)
             {
-                return $"XXXX-WXX-{timex.DayOfWeek}";
+                return $"{Constants.TimexFuzzyYear}-{Constants.TimexFuzzyWeek}-{timex.DayOfWeek}";
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs:139]: Got it, I will modify it.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/Constants.cs [hunk 5]
```diff
@@ -33,6 +33,14 @@ public static class Constants
         public const string HourUnit = "hour";
         public const string MinuteUnit = "minute";
         public const string SecondUnit = "second";
+        public const string TimeDurationUnit = "s";
+
+        public const string Now = "now";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/Constants.cs:38]: Are these general terms? "now" and "every"? If English specific, they should be resources or constants in an English file.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs [hunk 6]
```diff
@@ -64,47 +65,43 @@ public static string Format(TimexProperty timex)
 
         private static string FormatDuration(TimexProperty timex)
         {
-            var isTimeDurationAlreadyExist = false;
-            var timexBuilder = new StringBuilder(Constants.GeneralPeriodPrefix);
+            var timexList = new List<string> { };
             if (timex.Years != null)
             {
-                timexBuilder.Append($"{timex.Years}Y");
+                timexList.Add($"{Constants.GeneralPeriodPrefix}{timex.Years}{Constants.TimexYear}");
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs:71]: All generation of timexes should come from methods that receive the parameters needed to create it.
Only that method should know which string to use.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs [hunk 7]
```diff
@@ -124,22 +121,7 @@ private static string FormatTime(TimexProperty timex)
...
-                return $"{Constants.TimexFuzzyYear}-{TimexDateHelpers.FixedFormatNumber(timex.Month, 2)}-{TimexDateHelpers.FixedFormatNumber(timex.DayOfMonth, 2)}";
-            }
-
-            if (timex.DayOfWeek != null)
-            {
-                return $"{Constants.TimexFuzzyYear}-{Constants.TimexFuzzyWeek}-{timex.DayOfWeek}";
-            }
-
-            return string.Empty;
+            return TimexHelpers.GenerateDateTimex(timex.Year ?? -1, timex.Month ?? -1, timex.DayOfWeek != null ? timex.DayOfWeek.Value : timex.DayOfMonth ?? -1, timex.DayOfWeek != null);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexFormat.cs:124]: If -1 is supposed to mean some invalid value, that should be a constant.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexHelpers.cs [hunk 8]
```diff
@@ -254,6 +257,46 @@ public static TimeRange TimeRangeFromTimex(TimexProperty timex)
...
+                    timexBuilder.Append($"{Constants.TimeTimexPrefix}{GetDurationTimexWithoutPrefix(timexComponent)}");
+                    isTimeDurationAlreadyExist = true;
+                }
+                else
+                {
+                    timexBuilder.Append($"{GetDurationTimexWithoutPrefix(timexComponent)}");
+                }
+            }
+
+            return timexBuilder.ToString();
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexHelpers.cs:279]: Invariant culture here.

**Review by tellarin — CHANGES_REQUESTED** [review]

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexHelpers.cs (hunk 8 above)
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexHelpers.cs:330]: stringBuilder.Tostring has no IFormatProvider parameter, can I use AppendFormat to assure Ivariant Culture instead?

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexHelpers.cs:330]: Interesting. Leave this one as-is then.

File: .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs [hunk 9]
```diff
@@ -328,7 +328,7 @@ public void DataTypes_Resolver_DateRange_Demaical_Period_PT()
             var today = new System.DateTime(2019, 4, 30);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs:328]: Test method name should be: DataTypes_Resolver_DateRange_Decimal_Period_PT()

**Review by tellarin — APPROVED** [review]
The minor rename can come in a future PR. Thanks.

**Pull request merged**